internal/hooks/v0hooks/v0hooks.go

```diff
@@ -15,13 +19,76 @@ const (
 	CustomizeAccessToken Name = "customize-access-token"
 	MFAVerification      Name = "mfa-verification"
 	PasswordVerification Name = "password-verification"
+	BeforeUserCreated    Name = "before-user-created"
+	AfterUserCreated     Name = "after-user-created"
 )
 
-// Hook Names
 const (
 	HookRejection = "reject"
 )
 
+const (
+	DefaultMFAHookRejectionMessage      = "Further MFA verification attempts will be rejected."
+	DefaultPasswordHookRejectionMessage = "Further password verification attempts will be rejected."
+)
+
+type Metadata struct {
+	UUID uuid.UUID `json:"uuid"`
+	Time time.Time `json:"time"`
+
+	// Hook name
+	Name Name `json:"name,omitempty"`
+
+	// IP Address of the request, if present
+	IPAddress string `json:"ip_address,omitempty"`
+}
+
+func NewMetadata(r *http.Request, name Name) *Metadata {
+	return &Metadata{
+		UUID:      uuid.Must(uuid.NewV4()),
+		Time:      time.Now(),
+		IPAddress: utilities.GetIPAddress(r),
+		Name:      name,
+	}
+}
+
+type BeforeUserCreatedInput struct {
+	Metadata *Metadata    `json:"metadata"`
+	User     *models.User `json:"user"`
+}
+
+func NewBeforeUserCreatedInput(
+	r *http.Request,
+	user *models.User,
+) *BeforeUserCreatedInput {
+	return &BeforeUserCreatedInput{
+		Metadata: NewMetadata(r, BeforeUserCreated),
+		User:     user,
+	}
+}
+
+type BeforeUserCreatedOutput struct {
+	Decision string `json:"decision"`
+	Message  string `json:"message"`
+}
+
+type AfterUserCreatedInput struct {
+	Metadata *Metadata    `json:"metadata"`
+	User     *models.User `json:"user"`
+}
+
+func NewAfterUserCreatedInput(
+	r *http.Request,
+	user *models.User,
+) *AfterUserCreatedInput {
+	return &AfterUserCreatedInput{
+		Metadata: NewMetadata(r, AfterUserCreated),
+		User:     user,
+	}
+}
```

Comment:
Can't we just use the zero struct values, cutting down on more code paths to test?

Reply:
Well the zero values are not valid, the main purpose of the New is to ensure the [Metadata](https://github.com/supabase/auth/pull/2028/files/ca67be0db26bd096697de8b0b53346791f3a3268#diff-070d7377dd35255e1da96c9252a425b4f731c54d85f3c7675bd3a3cff522e5c3R46) is setup. Without them every single place we invoke a hook will look like:
```Go
	input := &BeforeUserCreatedInput{
		Metadata: &Metadata{
			UUID:      uuid.Must(uuid.NewV4()),
			Time:      time.Now(),
			IPAddress: utilities.GetIPAddress(r),
			Name:      BeforeUserCreated,
		},
		User: user,
	}
```

Reply:
Or...

```go
	input := &BeforeUserCreatedInput{
		Metadata: NewMetadataFromRequest(r, BeforeUserCreated),
		User: user,
	}
```

Reply:
Actually in general I prefer if the `Metadata` is set as part of `InvokeHook` so callers only fill in the relevant payload.

Reply:
New* funcs to create objects is an established pattern in Go and we use it all over the code base, are we sure we want to [debate these minor points](https://en.wikipedia.org/wiki/Law_of_triviality)?